compiler/rustc_mir_transform/src/lib.rs

```diff
@@ -592,6 +593,14 @@ fn inner_optimized_mir(tcx: TyCtxt<'_>, did: LocalDefId) -> Body<'_> {
         None => {}
         Some(other) => panic!("do not use `optimized_mir` for constants: {:?}", other),
     }
+
+    // Run `mir_inliner_callers_count_map` query for inliner
+    // which depends on `mir_drops_elaborated_and_const_checked`
+    // which we are going to steal below, so cache it's result.
+    if crate::inline::is_enabled(tcx) {
+        let _ = tcx.mir_inliner_callers_count_map(());
```

Comment:
`mir_inliner_callers_count_map` is whole-crate analysis. By calling it here, the computation of `optimized_mir` depends on this whole-crate analysis, and needs to be recomputed each time *any* caller count is changed in the crate. This dependency needs to be threaded through a `mir_inliner_callers_count(def_id)` query to avoid this invalidation.

Reply:
I'm not sure, what kind of dependency should be passed here. 

I mean `optimized_mir` really starts to depend on the number of calls in the functions, but that's the whole point of this PR. 

I can try to add per `def_id` computation of the number of callers, but it will turn algorithm from O(N) to O(N^2), because it will iterate over crate for every function in the crate.

I can add this add additional indirection `mir_inliner_callers_count(def_id) -> mir_inliner_callers_count_map(()) -> ...`, but will it give any difference? 

I mean `mir_inliner_callers_count(def_id)` will be still invalidated, when any caller count is changed.

I can add one more additional layer to break the dependency and recompute only what's needed.
`mir_inliner_crate_callers_count(def_id) -> mir_keys() -> mir_inliner_callers_count(key_def_id, def_id) `

Would you prefer the last option?

Reply:
I agree that we want to modify the optimization of a function `f` depending on the number of calls to `f`. I want to avoid having to redo the optimization because of a change to the number of calls to `g`.

I mean I'd rather have `optimized_mir(def_id) -> mir_inliner_callers_count(def_id) -> mir_inliner_callers_count_map(())`. That way, if `mir_inliner_callers_count_map` changes for an unrelated `def_id2`, the query system will recompute `mir_inliner_callers_count(def_id)`, which is cheap. The query system will see that `mir_inliner_callers_count(def_id)` has not changed. Therefore `optimized_mir` won't be recomputed.